### PR TITLE
npmlog: Add missing functions by extending EventEmitter

### PR DESCRIPTION
**Pull request by wildlyinaccurate**
Hello! This change is slightly more substantial than I hoped it would be. The [object exported by npmlog](https://github.com/npm/npmlog/blob/master/log.js#L4-L5) is an instance of EventEmitter, and I wanted this to be reflected in the types.

The only thing I'm not happy about with this PR is my use of `any`. I'm not proficient enough to be able to come up with a type that dynamically adds properties to the npmlog object. If anyone has advice on this, I'd be happy to update the PR.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/npm/npmlog/blob/master/log.js#L4-L5
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/npmlog/index.d.ts [hunk 1]
```diff
@@ -2,65 +2,81 @@
...
-    level: string;
-    prefix: string;
-    message: string;
-    messageRaw: string;
-}
+        log(level: LogLevels | string, message: string, ...args: any[]): void;
+        log(level: LogLevels | string, prefix: string, message: string, ...args: any[]): void;
 
-// TODO: newStream, newGroup, setGaugeTemplate and setGaugeTemplateSet need to be added
+        silly(message: string, ...args: any[]): void;
```
Comment (littlepiggy03) [inline types/npmlog/index.d.ts]: README of `npmlog` didn't mention the usage ignoring prefix. (https://github.com/npm/npmlog#basic-usage)

Screenshot as another proof: (already confirmed the version is newest)

the screenshot

File: types/npmlog/npmlog-tests.ts [hunk 2]
```diff
@@ -1,9 +1,11 @@
 import * as npmlog from "npmlog";
+import { MessageObject } from "npmlog";
 
 const prefix = "str";
 const message = "otherStr";
 
 ['silly', 'verbose', 'info', 'http', 'warn', 'error'].forEach(lvl => npmlog.log(lvl, prefix, message));
+['silly', 'verbose', 'info', 'http', 'warn', 'error'].forEach(lvl => npmlog.log(lvl, message));
```
Comment (littlepiggy03) [inline types/npmlog/npmlog-tests.ts]: Same as above.

File: types/npmlog/index.d.ts [hunk 3]
```diff
@@ -2,65 +2,81 @@
 // Project: https://github.com/npm/npmlog#readme
 // Definitions by: Daniel Schmidt <https://github.com/DanielMSchmidt>
 //                 Zhu Zijia <https://github.com/littlepiggy03>
+//                 Joseph Wynn <https://github.com/wildlyinaccurate>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.4
+// TypeScript Version: 3.4
```
Comment (littlepiggy03) [inline types/npmlog/index.d.ts]: Probably this is the minimal TypeScript version required for compiling this type definition, so don't need to edit it? (I'm not so sure.)

(See https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-use-features-from-typescript-21-or-above and https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-use-features-from-typescript-31-or-above)

**Review by ghost — CHANGES_REQUESTED** [review]
Nice work! But there are some tiny issues need to be resolved.

**Comment by wildlyinaccurate** [comment]
@littlepiggy03 thanks for the review! I totally misread the npmlog documentation. It says:

> prefix {String} A string prefix. Set to "" to skip.

I will revert my change to remove `prefix` and also revert the TS version

**Comment by wildlyinaccurate** [comment]
I've also added some missing log levels.

**Review by ghost — APPROVED** [review]
I think it can be merged now. Thanks for your contribution!

File: types/npmlog/tslint.json [hunk 4]
```diff
@@ -1 +1,11 @@
-{ "extends": "dtslint/dt.json" }
+{
+    "extends": "dtslint/dt.json",
+    "rules": {
```
Comment (sheetalkamat) [inline types/npmlog/tslint.json]: Please remove the rules here and fix errors. We want adherence to lint rules.

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/npmlog/tslint.json (hunk 4 above)
Comment (wildlyinaccurate) [reply types/npmlog/tslint.json]: @sheetalkamat Can you advise how I can adhere to this rule in npmlog-tests.ts? I can't figure out the syntax to combine those into a single import.

File: types/npmlog/npmlog-tests.ts [hunk 5]
```diff
@@ -42,3 +44,20 @@ npmlog.addLevel("styled-level", 42, {
...
 }, 'display name');
+
+npmlog.addLevel("broadcast", 10);
+
+npmlog.broadcast(prefix, message);
+npmlog.broadcast(message);
+
+npmlog.on("broadcast", () => {});
+
+const msg: MessageObject = {
```
Comment (sheetalkamat) [inline types/npmlog/npmlog-tests.ts]: Fix error by writing npmlog.MessageObject here instead of adding another import?

Comment (wildlyinaccurate) [reply types/npmlog/npmlog-tests.ts]: Thanks - done!

**Comment by wildlyinaccurate** [comment]
@sheetalkamat are you happy with the latest changes?

**Comment by wildlyinaccurate** [comment]
@sheetalkamat sorry to keep pinging. You requested changes a couple of weeks ago. I would be grateful if you could review my latest changes so I can progress with this PR.

**Comment by wildlyinaccurate** [comment]
Merged in #36417.

**Pull request closed without merging**